types/sharded_map.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * SPDX-FileCopyrightText: © Hypermode Inc. <hello@hypermode.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Package types contains some very common utilities used by Dgraph. These utilities
+// are of "miscellaneous" nature, e.g., error checking.
+package types
+
+import (
+	"sync"
+)
+
+type ShardedMap struct {
+	Shards []map[uint64]Val
+}
+
+const NumShards = 30
+
+func NewShardedMap() *ShardedMap {
+	shards := make([]map[uint64]Val, NumShards)
+	for i := range shards {
+		shards[i] = make(map[uint64]Val)
+	}
+	return &ShardedMap{Shards: shards}
+}
+
+func (s *ShardedMap) Merge(other *ShardedMap, ag func(a, b Val) Val) {
+	var wg sync.WaitGroup
+	for i := range s.Shards {
+		wg.Add(1)
+		go func(i int) {
+			for k, v := range s.Shards[i] {
+				val, ok := other.Shards[i][k]
+				if !ok {
+					continue
+				}
+				s.Shards[i][k] = ag(val, v)
+			}
+			for k, v := range other.Shards[i] {
```

Comment:
can we not do just one iteration over other.Shards[i] and merge if found in s.Shards[i] otherwise, write in s.Shards[i]